cdc/owner/scheduler.go

```diff
@@ -0,0 +1,382 @@
+	s.moveTableJobQueue = append(s.moveTableJobQueue, &moveTableJob{
+		tableID: tableID,
+		target:  target,
+	})
+}
+
+func (s *scheduler) handleMoveTableJob() (shouldUpdateState bool, err error) {
+	shouldUpdateState = true
+	if len(s.moveTableJobQueue) == 0 {
+		return
+	}
+	table2CaptureIndex, err := s.table2CaptureIndex()
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+	for _, job := range s.moveTableJobQueue {
+		source, exist := table2CaptureIndex[job.tableID]
+		if !exist {
+			return
+		}
+		s.moveTableTargets[job.tableID] = job.target
+		job := job
+		shouldUpdateState = false
+		s.state.PatchTaskStatus(source, func(status *model.TaskStatus) (*model.TaskStatus, bool, error) {
+			if status == nil {
+				// the capture may be down, just skip remove this table
+				return status, false, nil
+			}
+			if status.Operation != nil && status.Operation[job.tableID] != nil {
+				// skip removing this table to avoid the remove operation created by the rebalance function interfering with the operation created by another function
+				return status, false, nil
+			}
+			status.RemoveTable(job.tableID, s.state.Status.CheckpointTs, false)
+			return status, true, nil
+		})
+	}
+	s.moveTableJobQueue = nil
+	return
+}
+
+func (s *scheduler) Rebalance() {
+	s.needRebalanceNextTick = true
+}
+
+func (s *scheduler) table2CaptureIndex() (map[model.TableID]model.CaptureID, error) {
+	table2CaptureIndex := make(map[model.TableID]model.CaptureID)
+	for captureID, taskStatus := range s.state.TaskStatuses {
+		for tableID := range taskStatus.Tables {
+			if preCaptureID, exist := table2CaptureIndex[tableID]; exist && preCaptureID != captureID {
+				return nil, cerror.ErrTableListenReplicated.GenWithStackByArgs(preCaptureID, captureID)
+			}
+			table2CaptureIndex[tableID] = captureID
+		}
+		for tableID := range taskStatus.Operation {
+			if preCaptureID, exist := table2CaptureIndex[tableID]; exist && preCaptureID != captureID {
+				return nil, cerror.ErrTableListenReplicated.GenWithStackByArgs(preCaptureID, captureID)
+			}
+			table2CaptureIndex[tableID] = captureID
+		}
+	}
+	return table2CaptureIndex, nil
+}
+
+// dispatchToTargetCaptures sets the the TargetCapture of scheduler jobs
+// If the TargetCapture of a job is not set, it chooses a capture with the minimum workload and sets the TargetCapture to the capture.
+func (s *scheduler) dispatchToTargetCaptures(pendingJobs []*schedulerJob) {
+	workloads := make(map[model.CaptureID]uint64)
+
+	for captureID := range s.captures {
+		workloads[captureID] = 0
+		taskWorkload := s.state.Workloads[captureID]
+		if taskWorkload == nil {
+			continue
+		}
+		for _, workload := range taskWorkload {
+			workloads[captureID] += workload.Workload
+		}
+	}
+
+	for _, pendingJob := range pendingJobs {
+		if pendingJob.TargetCapture == "" {
+			target, exist := s.moveTableTargets[pendingJob.TableID]
+			if !exist {
+				continue
+			}
```

Comment:
If the `TargetCapture` of a `pendingJob` is not set, and this table is not in `moveTableTargets`, how and where is the candidate target capture chosen?

Reply:
`TargetCapture` is set by L218